pkg/updatestatus/controlplaneinformer.go

```diff
@@ -64,31 +74,189 @@ func newControlPlaneInformerController(
+	}
+
+	var available *configv1.ClusterOperatorStatusCondition
+	var degraded *configv1.ClusterOperatorStatusCondition
+	var progressing *configv1.ClusterOperatorStatusCondition
+
+	for _, condition := range operator.Status.Conditions {
+		condition := condition
+		switch {
+		case condition.Type == configv1.OperatorAvailable:
+			available = &condition
+		case condition.Type == configv1.OperatorDegraded:
+			degraded = &condition
+		case condition.Type == configv1.OperatorProgressing:
+			progressing = &condition
+		}
+	}
+
+	if progressing != nil {
+		if !updated && progressing.Status == configv1.ConditionTrue {
+			updating.Status = metav1.ConditionTrue
+			updating.Reason = string(ClusterOperatorUpdatingReasonProgressing)
+			updating.Message = progressing.Message
+		}
+		if !updated && progressing.Status == configv1.ConditionFalse {
+			updating.Status = metav1.ConditionFalse
+			updating.Reason = string(ClusterOperatorUpdatingReasonPending)
+			updating.Message = progressing.Message
+		}
+	}
+
+	health := metav1.Condition{
+		Type:               string(ClusterOperatorStatusInsightHealthy),
+		Status:             metav1.ConditionTrue,
+		Reason:             string(ClusterOperatorHealthyReasonAsExpected),
+		LastTransitionTime: now,
+	}
+
+	if available == nil {
+		health.Status = metav1.ConditionUnknown
+		health.Reason = string(ClusterOperatorHealthyReasonUnavailable)
+		health.Message = "The cluster operator is unavailable because the available condition is not found in the cluster operator's status"
+	} else if available.Status != configv1.ConditionTrue {
+		health.Status = metav1.ConditionFalse
+		health.Reason = string(ClusterOperatorHealthyReasonUnavailable)
+		health.Message = available.Message
+	} else if degraded != nil && degraded.Status == configv1.ConditionTrue {
+		health.Status = metav1.ConditionFalse
+		health.Reason = string(ClusterOperatorHealthyReasonDegraded)
+		health.Message = degraded.Message
+	}
+
+	return &ClusterOperatorStatusInsight{
+		Name: operator.Name,
+		Resource: ResourceRef{
+			Resource: "clusteroperators",
+			Group:    configv1.GroupName,
+			Name:     operator.Name,
+		},
+		Conditions: []metav1.Condition{updating, health},
+	}
+}
+
+func getImagePullSpec(ctx context.Context, name string, appsClient appsv1client.AppsV1Interface) string {
+	// It is known that the image pull spec for co/machine-config can be accessed from the deployment
+	if name == "machine-config" && appsClient != nil {
+		mcoDeployment, err := appsClient.Deployments("openshift-machine-config-operator").Get(ctx, "machine-config-operator", metav1.GetOptions{})
+		if err != nil {
+			klog.Warningf("error getting machine-config-operator deployment: %v", err)
+		}
+		for _, c := range mcoDeployment.Spec.Template.Spec.Containers {
+			if c.Name == "machine-config-operator" {
+				return c.Image
+			}
+		}
+	}
+	// We may add here retrieval of the image pull spec for other COs when they implemented "operator-image" in the status.versions
```

Comment:
OH! So we are not special-casing MCO like I thought. Now I understand. We really need special treatment for any CO that chooses to respect the `operator-image` version idea.

I think we will need to address this in the future but let's not bog down on this. We can do it in a followup. I think we could do something like:
1. Check if the CO set a non-empty `operator-image`
2. If yes, iterate over Deployments it owns, then iterate over their containers, and if we discover the value from (1), consider it "updated". That would allow us to not need explicit namespace/name/container name values?

The above is just me thinking loud, I don't think we need to address it here.

Reply:
I checked out the algorithm above with `co/machine-config`:

```console
$ oc get deployment -n openshift-machine-config-operator machine-config-operator -o yaml | yq -y .metadata.ownerReferences
- apiVersion: config.openshift.io/v1
  controller: true
  kind: ClusterVersion
  name: version
  uid: 263cb902-78c7-4fec-8254-161775dd3ee5
```

The owner is our `cv/version`, instead of `co/machine-config`. But it would be wonderful if could make an algorithm like it.